src/Compilers/VisualBasic/Portable/Syntax/SyntaxFactory.vb

```diff
@@ -532,5 +532,91 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+                node = node.Parent
+            Loop
+
+            Return False
+        End Function
+
+        ''' <summary>
+        ''' Determines if a submission is complete.
+        ''' Returns false if the syntax is valid but incomplete.
+        ''' Returns true if the syntax is invalid or complete.
+        ''' </summary>
+        ''' <param name="tree">Syntax tree.</param>
+        Public Shared Function IsCompleteSubmission(tree As SyntaxTree) As Boolean
+            Dim options As VisualBasicParseOptions = DirectCast(tree.Options, VisualBasicParseOptions)
+            Dim languageVersion As LanguageVersion = options.LanguageVersion
+
+            If (tree Is Nothing) Then
+                Throw New ArgumentNullException(NameOf(tree))
+            End If
+
+            If (Not tree.HasCompilationUnitRoot) Then
+                Return False
+            End If
+
+            Dim compilation As CompilationUnitSyntax = DirectCast(tree.GetRoot(), CompilationUnitSyntax)
+
+            For Each err In compilation.GetDiagnostics()
+                Select Case DirectCast(err.Code, ERRID)
+                    Case ERRID.ERR_LbExpectedEndIf,
+                         ERRID.ERR_ExpectedEndRegion
+                        Return False
+                    Case ERRID.ERR_ExpectedEOS
+                        ' Invalid statements that should have been separated by a newline or a colon.
+                        ' For example: `If condition statement`
+                        Return True
+                End Select
+            Next
+
+            Dim lastNode = compilation.ChildNodes().LastOrDefault()
+            If lastNode Is Nothing Then
+                ' Invalid submission. The compilation does not have any children.
+                Return True
+            End If
+
+            Dim lastToken = lastNode.GetLastToken(includeZeroWidth:=True, includeSkipped:=True)
+
+            If IsPartOfUnterminatedLinqQuery(lastToken, lastNode, compilation.EndOfFileToken) OrElse
+                    (lastToken.HasTrailingTrivia AndAlso lastToken.TrailingTrivia.Last().IsKind(SyntaxKind.LineContinuationTrivia)) Then
+                ' Even if the compilation is correct but has a line continuation trivia return statement as incomplete.
+                ' For example `Dim x = 12 _` has no compilation errors but should be treated as an incomplete statement.
+                Return False
+            ElseIf Not compilation.HasErrors Then
+                ' No errors returned. This is a valid and complete submission.
+                Return True
+            ElseIf lastNode.IsKind(SyntaxKind.IncompleteMember) OrElse lastToken.IsMissing Then
+                Return False
+            End If
+
+            For Each err In lastToken.GetDiagnostics()
+                Select Case DirectCast(err.Code, ERRID)
+                    Case ERRID.ERR_UnterminatedStringLiteral
+                        If languageVersion = LanguageVersion.VisualBasic14 Then
```

Comment:
Should this be an equality comparison? What about the future language versions? Please add a comment explaining why language version is significant.